File: MATLAB/example_6/example_6.m
% %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% %
%                                                                         %
%   EEEEEE  XX  XX   AAAA   MM   MM  PPPPPP  LL      EEEEEE      666666   %
%   EE       XXXX   AA  AA  MMM MMM  PP  PP  LL      EE          66       %
%   EEEEE     XX    AA  AA  MMMMMMM  PPPPPP  LL      EEEEE       666666   %
%   EE       XXXX   AAAAAA  MM   MM  PP      LL      EE          66  66   %
%   EEEEEE  XX  XX  AA  AA  MM   MM  PP      LLLLLL  EEEEEE      666666   %
%                                                                         %
% %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% %

% Example from the following paper
%  Li, G., and H. Rabitz (2012), General formulation of HDMR component 
%      functions with independent and correlated variables, J. 
%      Math. Chem., 50, pp. 99-130

clc; clear; close all hidden;               % clear workspace and figures

d = 3;                                      % # parameters? 
N = 5000;                                   % # samples to be used
mu = [0.5 0.5 0.5];                         % Sample means, µ
s = [0.2 0.2 0.18];                         % Sample stds, σ
Cov = @(s,r12) [ s(1)^2 r12*s(1)*s(2) 0 ; ...
        r12*s(1)*s(2) s(2)^2 0 ; 0 0 s(3)^2 ];
r12 = 0.6; C = Cov(s,r12);                  % Create dxd covariance matrix, Σ
X = mvnrnd(mu,C,N);                         % Draw N samples from N(µ,Σ) 
X = (X - repmat(min(X),N,1))./ ...          % Normalize X - values
    repmat(max(X)-min(X),N,1);
a=[1 2]; b=[2 3]; c=[3 1 2]; e=[1 2 2 3];   % Inputs of user-defined function
y = user_function(X,mu,a,b,c,e);            % Evaluate function y = f(X)
sigma2 = var(y)/100;                        % Variance of random error
y = y + normrnd(0,sqrt(sigma2),N,1);        % Add random error, y = Nx1 vector

options = struct('graphics',1, ...          % Specify HDMR options
    'maxorder',3,'maxiter',100, ...
    'bf1',1,'bf2',0,'bf3',0,...
    'm',2,'K',10,'R',300,'method',1, ...
    'alfa',0.01,'lambda',0.10, ...
    'vartol',1e-3,'refit',1);
[S,Ss,Fx,Em,Xy] = HDMR(X,y,options);        % Now run the HDMR toolbox

% options = struct('graphics',1,'maxorder',3,'maxiter',100, ...
%     'm',2,'K',10,'R',300,'method',2,'alfa',0.99, ...
%     'lambda',10,'vartol',1e-3,'refit',1);

% subplot(2,2,1),plot(HDMR.X_n(:,1),Y_ebf(:,1),'b.');
% xlabel('x1','fontsize',16); ylabel('f1(x1)','fontsize',16); 
% set(gca,'fontsize',16);
% subplot(2,2,2),plot(HDMR.X_n(:,2),Y_ebf(:,2),'b.');
% xlabel('x2','fontsize',16); ylabel('f2(x2)','fontsize',16); 
% set(gca,'fontsize',16);
% subplot(2,2,3),plot(HDMR.X_n(:,3),Y_ebf(:,3),'b.');
% xlabel('x3','fontsize',16); ylabel('f3(x3)','fontsize',16); 
% set(gca,'fontsize',16);
% subplot(2,2,4),plot(X(:,3),Y_ebf(:,3),'b');

File: Python/example_6/example_6.py
# ####################################################################### #
#                                                                         #
#   EEEEEE  XX  XX   AAAA   MM   MM  PPPPPP  LL      EEEEEE      666666   #
#   EE       XXXX   AA  AA  MMM MMM  PP  PP  LL      EE          66       #
#   EEEEE     XX    AA  AA  MMMMMMM  PPPPPP  LL      EEEEE       666666   #
#   EE       XXXX   AAAAAA  MM   MM  PP      LL      EE          66  66   #
#   EEEEEE  XX  XX  AA  AA  MM   MM  PP      LLLLLL  EEEEEE      666666   #
#                                                                         #
# ####################################################################### #

# Example from the following paper
#  Li, G., and H. Rabitz (2012), General formulation of HDMR component 
#      functions with independent and correlated variables, J. 
#      Math. Chem., 50, pp. 99-130

import sys
import os
import numpy as np

# Get the current working directory
current_directory = os.getcwd()
# Go up one directory
parent_directory = os.path.abspath(os.path.join(current_directory, '..'))
# add this to path
sys.path.append(parent_directory)
# Add another directory
misc_directory = os.path.abspath(os.path.join(parent_directory, 'miscellaneous'))
# add this to path
sys.path.append(misc_directory)

from HDMR import HDMR
from user_function import *

# Case study
d = 3           		# Number of parameters
N = 5000        		# Number of samples
mu = np.array([0.5, 0.5, 0.5])  # Mean values for the distribution
s = np.array([0.2, 0.2, 0.18])  # Standard deviations
r12 = 0.6       		# Correlation between the first two variables

# Create covariance matrix
C = create_covariance(s, r12)

# Draw N samples from a multivariate normal distribution
X = np.random.multivariate_normal(mu, C, N)

# Normalize X
X_min = X.min(axis=0)
X_max = X.max(axis=0)
X = (X - X_min) / (X_max - X_min)

# Define user function parameters
a = np.array([1, 2])
b = np.array([2, 3])
c = np.array([3, 1, 2])
e = np.array([1, 2, 2, 3])

# Evaluate user-defined function y = f(X)
y = user_function(X, mu, a, b, c, e)

# Add random noise
sigma2 = np.var(y) / 100
noise = np.random.normal(0, np.sqrt(sigma2), N)
y = y + noise

# Specify HDMR options (assuming you have a Python implementation of HDMR)
options = {
    'graphics': 1,
    'maxorder': 3,
    'maxiter': 100,
    'bf1': 1,
    'bf2': 0,
    'bf3': 0,
    'm': 2,
    'K': 10,
    'R': 300,
    'method': 1,
    'alfa': 0.01,
    'lambda': 0.10,
    'vartol': 1e-3,
    'refit': 1
}

if __name__ == '__main__':
	# Run the HDMR toolbox
	S, Ss, Fx, Em, Xy, RT = HDMR(X, y, options)

	print("y (with random noise added):", y)
